equistore-core/src/blocks.rs

```diff
@@ -104,152 +95,91 @@ fn check_component_labels(components: &[Arc<Labels>]) -> Result<(), Error> {
-            "unable to find [{}] in the components ", dimensions.join(", ")
-        )))?;
-
-        let moved_component = self.components.0.remove(component_axis);
-
-        // construct the new property with old properties and the components
-        let old_properties = &self.properties;
-        let new_property_names = moved_component.names().iter()
-            .chain(old_properties.names().iter())
-            .copied()
-            .collect();
-
-        let mut new_properties_builder = LabelsBuilder::new(new_property_names);
-        for new_property in moved_component.iter() {
-            for old_property in old_properties.iter() {
-                let mut property = new_property.to_vec();
-                property.extend_from_slice(old_property);
-                new_properties_builder.add(&property)?;
-            }
-        }
-        let new_properties = new_properties_builder.finish();
-
-        let mut new_shape = self.data.shape()?.to_vec();
-        let properties_axis = new_shape.len() - 1;
-        new_shape[properties_axis] = new_properties.count();
-        new_shape.remove(component_axis + 1);
-
-        self.data.swap_axes(component_axis + 1, properties_axis - 1)?;
-        self.data.reshape(&new_shape)?;
-
-        self.properties = Arc::new(new_properties);
-
-        Ok(())
-    }
-
-    /// Try to copy this `BasicBlock`. This can fail if we are unable to copy
-    /// the underlying `eqs_array_t` data array
-    pub fn try_clone(&self) -> Result<BasicBlock, Error> {
-        let data = self.data.try_clone()?;
-
-        Ok(BasicBlock {
-            data,
-            samples: Arc::clone(&self.samples),
-            components: self.components.clone(),
-            properties: Arc::clone(&self.properties),
-        })
-    }
-}
-
 /// A single block in a `TensorMap`, containing both values & optionally
 /// gradients of these values w.r.t. any relevant quantity.
 #[derive(Debug)]
 pub struct TensorBlock {
-    values: BasicBlock,
-    gradients: HashMap<String, BasicBlock>,
+    pub values: eqs_array_t,
+    pub samples: Arc<Labels>,
+    pub components: ImmutableVec<Arc<Labels>>,
+    pub properties: Arc<Labels>,
+    gradients: HashMap<String, TensorBlock>,
     // all the keys from `self.gradients`, as C-compatible strings
     gradient_parameters: Vec<ConstCString>,
 }
 
 impl TensorBlock {
-    /// Create a new `TensorBlock` containing the given data, described by the
+    /// Create a new `TensorBlock` containing the given values, described by the
     /// `samples`, `components`, and `properties` labels. The block is
     /// initialized without any gradients.
     pub fn new(
-        data: impl Into<eqs_array_t>,
+        values: eqs_array_t,
         samples: Arc<Labels>,
         components: Vec<Arc<Labels>>,
         properties: Arc<Labels>,
     ) -> Result<TensorBlock, Error> {
+        check_values_and_labels(
+            "values and labels don't match", &values, &samples, &components, &properties
```

Comment:
```suggestion
            "data and labels don't match", &values, &samples, &components, &properties
```

I would prefer to keep "data" in the error messages, since we are still talking about "data & metadata". The fact that you access the data through the `values` attribute feels unrelated